File: src/graph/retriever.js
import { join } from "node:path";
import { readText } from "../core/files.js";

export function buildIssueQuery(issue, intent = {}) {
  return [
    issue.title,
    issue.body,
    intent.observedBehaviour,
    intent.expectedBehaviour,
    ...(intent.symbols ?? []),
    ...(intent.candidateFiles ?? [])
  ].filter(Boolean).join("\n");
}

export async function retrieveIssueContext({ graph, repoPath, issue, intent, maxFiles = 8 }) {
  const query = buildIssueQuery(issue, intent);
  const terms = tokenise(query);
  const scored = scoreNodes(graph, terms);
  const fileScores = new Map();

  for (const item of scored) {
    const path = item.node.path;
    if (!path) {
      continue;
    }
    fileScores.set(path, (fileScores.get(path) ?? 0) + item.score);
  }

  for (const file of intent?.candidateFiles ?? []) {
    fileScores.set(file, (fileScores.get(file) ?? 0) + 50);
  }

  const files = [...fileScores.entries()]
    .sort((left, right) => right[1] - left[1])
    .slice(0, maxFiles)
    .map(([path, score]) => ({ path, score }));

  const snippets = [];
  for (const file of files) {
    const content = await safeRead(join(repoPath, file.path));
    if (!content) {
      continue;
    }
    snippets.push({
      path: file.path,
      score: file.score,
      excerpt: makeExcerpt(content, terms)
    });
  }

  return {
    queryTerms: [...terms],
    files,
    snippets,
    graphNodes: scored.slice(0, 40).map(({ node, score }) => ({
      id: node.id,
      type: node.type,
      name: node.name,
      path: node.path,
      line: node.line,
      score
    }))
  };
}

function scoreNodes(graph, terms) {
  const scored = [];
  for (const node of graph.nodes ?? []) {
    const haystack = [
      node.id,
      node.type,
      node.name,
      node.path,
      JSON.stringify(node.metadata ?? {})
    ].filter(Boolean).join(" ").toLowerCase();
    let score = 0;
    for (const term of terms) {
      if (!term || term.length < 2) {
        continue;
      }
      if (haystack.includes(term)) {
        score += term.length > 8 ? 6 : 3;
      }
      if (node.name?.toLowerCase() === term) {
        score += 20;
      }
      if (node.path?.toLowerCase().includes(term)) {
        score += 5;
      }
    }
    if (score > 0) {
      scored.push({ node, score });
    }
  }
  return scored.sort((left, right) => right.score - left.score);
}

function tokenise(text) {
  return new Set(
    String(text ?? "")
      .replace(/https?:\/\/\S+/g, " ")
      .split(/[^A-Za-z0-9_./-]+/)
      .map((term) => term.trim().toLowerCase())
      .filter((term) => term.length >= 3)
      .flatMap((term) => splitTerm(term))
  );
}

function splitTerm(term) {
  const terms = [term];
  if (term.includes("_")) {
    terms.push(...term.split("_"));
  }
  if (term.includes("-")) {
    terms.push(...term.split("-"));
  }
  if (term.includes("/")) {
    terms.push(...term.split("/"));
  }
  return terms.filter((item) => item.length >= 3);
}

function makeExcerpt(content, terms) {
  const lines = content.split("\n");
  const matching = [];
  for (let index = 0; index < lines.length; index += 1) {
    const lower = lines[index].toLowerCase();
    if ([...terms].some((term) => lower.includes(term))) {
      matching.push(index);
    }
  }

  const selected = new Set();
  for (const index of matching.slice(0, 12)) {
    for (let line = Math.max(0, index - 3); line <= Math.min(lines.length - 1, index + 4); line += 1) {
      selected.add(line);
    }
  }

  const ordered = [...selected].sort((left, right) => left - right);
  if (!ordered.length) {
    return lines.slice(0, 80).map((line, index) => `${index + 1}: ${line}`).join("\n");
  }

  return ordered.map((line) => `${line + 1}: ${lines[line]}`).join("\n");
}

async function safeRead(path) {
  try {
    return await readText(path);
  } catch {
    return null;
  }
}

File: go_issue_agent/graph/retriever.py
from __future__ import annotations

from pathlib import Path

from go_issue_agent.core.files import read_text


def build_issue_query(issue: dict, intent: dict | None = None) -> str:
    intent = intent or {}
    parts = [
        issue.get("title"),
        issue.get("body"),
        intent.get("observedBehaviour"),
        intent.get("expectedBehaviour"),
        *(intent.get("symbols") or []),
        *(intent.get("candidateFiles") or []),
    ]
    return "\n".join(part for part in parts if part)


def retrieve_issue_context(graph: dict, repo_path: str, issue: dict, intent: dict, max_files: int = 8) -> dict:
    query = build_issue_query(issue, intent)
    terms = _tokenise(query)
    scored = _score_nodes(graph, terms)
    file_scores: dict[str, int] = {}

    for item in scored:
        path = item["node"].get("path")
        if path:
            file_scores[path] = file_scores.get(path, 0) + item["score"]

    for file in intent.get("candidateFiles") or []:
        file_scores[file] = file_scores.get(file, 0) + 50

    files = [{"path": path, "score": score} for path, score in sorted(file_scores.items(), key=lambda item: item[1], reverse=True)[:max_files]]
    snippets = []
    for file in files:
        full_path = Path(repo_path) / file["path"]
        if not full_path.exists():
            continue
        content = read_text(full_path)
        snippets.append({"path": file["path"], "score": file["score"], "excerpt": _make_excerpt(content, terms)})

    return {
        "queryTerms": sorted(terms),
        "files": files,
        "snippets": snippets,
        "graphNodes": [
            {
                "id": item["node"]["id"],
                "type": item["node"]["type"],
                "name": item["node"].get("name"),
                "path": item["node"].get("path"),
                "line": item["node"].get("line"),
                "score": item["score"],
            }
            for item in scored[:40]
        ],
    }


def _score_nodes(graph: dict, terms: set[str]) -> list[dict]:
    scored = []
    for node in graph.get("nodes", []):
        haystack = " ".join(
            str(value).lower()
            for value in [
                node.get("id"),
                node.get("type"),
                node.get("name"),
                node.get("path"),
                node.get("metadata"),
            ]
            if value is not None
        )
        score = 0
        for term in terms:
            if len(term) < 2:
                continue
            if term in haystack:
                score += 6 if len(term) > 8 else 3
            if str(node.get("name", "")).lower() == term:
                score += 20
            if term in str(node.get("path", "")).lower():
                score += 5
        if score > 0:
            scored.append({"node": node, "score": score})
    return sorted(scored, key=lambda item: item["score"], reverse=True)


def _tokenise(text: str) -> set[str]:
    terms = set()
    for raw in text.replace("http://", " ").replace("https://", " ").split():
        for part in raw.replace("\n", " ").split():
            for piece in _split_term(part.lower()):
                cleaned = "".join(ch for ch in piece if ch.isalnum() or ch in "_./-")
                if len(cleaned) >= 3:
                    terms.add(cleaned)
    return terms


def _split_term(term: str) -> list[str]:
    pieces = [term]
    for separator in ["_", "-", "/"]:
        expanded = []
        for piece in pieces:
            expanded.extend(piece.split(separator))
        pieces.extend(expanded)
    return [piece for piece in pieces if len(piece) >= 3]


def _make_excerpt(content: str, terms: set[str]) -> str:
    lines = content.splitlines()
    matching = []
    for index, line in enumerate(lines):
        lower = line.lower()
        if any(term in lower for term in terms):
            matching.append(index)

    selected = set()
    for index in matching[:12]:
        for line in range(max(0, index - 3), min(len(lines) - 1, index + 4) + 1):
            selected.add(line)

    ordered = sorted(selected)
    if not ordered:
        return "\n".join(f"{index + 1}: {line}" for index, line in enumerate(lines[:80]))
    return "\n".join(f"{line + 1}: {lines[line]}" for line in ordered)
